File: src/PlantBasedPizza.Recipes/applications/PlantBasedPizza.Recipes.Core/Entities/Recipe.cs
using System.Text.Json.Serialization;

namespace PlantBasedPizza.Recipes.Core.Entities;

public enum RecipeCategory
{
    Pizza,
    Sides,
    Drinks
}

public class Recipe
{
    private List<Ingredient> _ingredients;
        
    [JsonConstructor]
    private Recipe()
    {
        RecipeIdentifier = "";
        Name = "";
        _ingredients = new List<Ingredient>();
    }
        
    public Recipe(RecipeCategory category, string recipeIdentifier, string name, decimal price)
    {
        Category = category;
        RecipeIdentifier = recipeIdentifier;
        Name = name;
        Price = price;
        _ingredients = new List<Ingredient>();
    }
        
    [JsonPropertyName("recipeIdentifier")]
    public string RecipeIdentifier { get; private set; }
        
    [JsonPropertyName("category")]
    public RecipeCategory Category { get; private set; }
        
    [JsonPropertyName("name")]
    public string Name { get; private set; }
        
    [JsonPropertyName("price")]
    public decimal Price { get; private set; }

    [JsonPropertyName("ingredients")]
    public IReadOnlyCollection<Ingredient> Ingredients => _ingredients;

    public void AddIngredient(string name, int quantity)
    {
        if (_ingredients == null)
        {
            _ingredients = new List<Ingredient>();
        }
            
        _ingredients.Add(new Ingredient(name, quantity));
    }
}

File: src/PlantBasedPizza.Recipes.Java/src/core/src/main/java/com/recipe/core/Recipe.java
package com.recipe.core;

import jakarta.persistence.*;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "recipes")
public class Recipe {
    public Recipe()
    {
    }

    @Id
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    @Column(name = "name")
    private String name;
    @Column(name = "category")
    private String category;
    @Column(name = "price")
    private Double price;

    @Column(name = "order_count")
    private int orderCount;

    @OneToMany(mappedBy="recipe", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<Ingredient> ingredients = new ArrayList<>();

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Double getPrice() {
        return price;
    }

    public void setPrice(Double price) {
        this.price = price;
    }

    public List<Ingredient> getIngredients() { return ingredients; }

    public void setIngredients(List<Ingredient> ingredients) {
        this.ingredients = ingredients;
        ingredients.forEach(ingredient -> ingredient.setRecipe(this));
    }

    public void addIngredient(String name, Number quantity){
        Ingredient ingredient = new Ingredient();
        ingredient.setName(name);
        ingredient.setQuantity(quantity);
        ingredient.setRecipe(this);

        this.ingredients.add(ingredient);
    }

    public RecipeDTO asDto()
    {
        RecipeDTO dto = new RecipeDTO();
        dto.setId(this.id);
        dto.setName(this.name);
        dto.setPrice(this.price);
        dto.setCategory(this.category);
        dto.setIngredients(this.ingredients);

        return dto;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public int getOrderCount() {
        return orderCount;
    }

    public void setOrderCount(int orderCount) {
        this.orderCount = orderCount;
    }

    public void incrementOrderCount() {
        this.orderCount++;
    }
}
